helper/src/main/java/io/github/libxposed/helpers/BaseFinder.java

```diff
@@ -0,0 +1,254 @@
+        stream = stream.filter(predicate);
+        return (S) this;
+    }
+
+    /**
+     * Filter methods/constructors if they have public modifier.
+     *
+     * @return this
+     */
+    public final S filterPublic() {
+        return filter(ModifierHelper::isPublic);
+    }
+
+    /**
+     * Filter methods/constructors if they not have public modifier.
+     *
+     * @return this
+     */
+    public final S filterNonPublic() {
+        return filter(ModifierHelper::isNotPublic);
+    }
+
+    /**
+     * Filter methods/constructors if they have protected modifier.
+     *
+     * @return this
+     */
+    public final S filterProtected() {
+        return filter(ModifierHelper::isProtected);
+    }
+
+    /**
+     * Filter methods/constructors if they not have protected modifier.
+     *
+     * @return this
+     */
+    public final S filterNonProtected() {
+        return filter(ModifierHelper::isNotProtected);
+    }
+
+    /**
+     * Filter methods/constructors if they have private modifier.
+     *
+     * @return this
+     */
+    public final S filterPrivate() {
+        return filter(ModifierHelper::isPrivate);
+    }
+
+    /**
+     * Filter methods/constructors if they not have private modifier.
+     *
+     * @return this
+     */
+    public final S filterNonPrivate() {
+        return filter(ModifierHelper::isNotPrivate);
+    }
+
+    /**
+     * Filter methods/constructors if they have native modifier.
+     *
+     * @return this
+     */
+    public final S filterNative() {
+        return filter(ModifierHelper::isNative);
+    }
+
+    /**
+     * Filter methods/constructors if they have native modifier.
+     *
+     * @return this
+     */
+    public final S filterNonNative() {
+        return filter(ModifierHelper::isNotNative);
+    }
+
+    /**
+     * Filter methods/constructors if they are package private.
+     *
+     * @return this
+     */
+    public final S filterPackage() {
+        return filter(ModifierHelper::isPackage);
+    }
+
+    /**
+     * Filter methods/constructors if they are not package private.
+     *
+     * @return this
+     */
+    public final S filterNonPackage() {
+        return filter(ModifierHelper::isNotPackage);
+    }
+
+    /**
+     * Filter methods/constructors by their exception types(Full-matches, cannot be null)
+     *
+     * @param e exception types
+     * @return this
+     */
+    public final S filterByExceptionType(Class<?>... e) {
+        var set1 = Set.of(e);
+
+        return filter(m -> {
+            var set2 = Set.of(getExceptionTypes(m));
+            return set1.equals(set2);
+        });
+    }
+
+    /**
+     * Filter methods/constructors by parameter types, make sure length is same as target method parameter types length.
+     *
+     * @param parameterTypes parameter types, use null to skip check some parameters.
+     * @return this
+     */
+    public final S filterByParameterTypes(Class<?>... parameterTypes) {
```

Comment:
I still think this is not extendable. What if I want a method with two ints in its parameters? Indices-specific filtering is not extendable. How about filtering by predicate but we pre-define some useful predicates like full matcher, include maters and exclude matcher. Same for exception types.

Reply:
I think this is a reasonable demand, parameter types must full-match(or null if skip).
```java
var method = MethodFinder.from(String.class, null)
            .filterByParameterTypes(char[].class, null, int.class)
            .first();
System.out.println(method.getName());

var method2 = MethodFinder.from(String.class, null)
            .filterByParameterTypes(char[].class, int.class, int.class)
            .first();
System.out.println(method2.getName());
```
output:
```
valueOf
valueOf
````

Reply:
if you want use more condition, why not use `BaseFinder#filter(Predicate<T>)` ?